app/src/main/java/org/onionshare/android/tor/TorManager.kt

```diff
@@ -0,0 +1,148 @@
+package org.onionshare.android.tor
+
+import android.app.Application
+import android.app.Application.BIND_AUTO_CREATE
+import android.content.BroadcastReceiver
+import android.content.ComponentName
+import android.content.Context
+import android.content.Intent
+import android.content.IntentFilter
+import android.content.ServiceConnection
+import android.os.IBinder
+import androidx.core.content.ContextCompat.startForegroundService
+import net.freehaven.tor.control.TorControlCommands.EVENT_CIRCUIT_STATUS
+import net.freehaven.tor.control.TorControlCommands.EVENT_ERR_MSG
+import net.freehaven.tor.control.TorControlCommands.EVENT_HS_DESC
+import net.freehaven.tor.control.TorControlCommands.EVENT_NEW_DESC
+import net.freehaven.tor.control.TorControlCommands.EVENT_OR_CONN_STATUS
+import net.freehaven.tor.control.TorControlCommands.EVENT_WARN_MSG
+import net.freehaven.tor.control.TorControlCommands.HS_ADDRESS
+import net.freehaven.tor.control.TorControlConnection
+import org.slf4j.LoggerFactory.getLogger
+import org.torproject.jni.TorService
+import org.torproject.jni.TorService.ACTION_STATUS
+import org.torproject.jni.TorService.EXTRA_STATUS
+import java.io.IOException
+import java.util.Collections
+import javax.inject.Inject
+import javax.inject.Singleton
+import kotlin.coroutines.Continuation
+import kotlin.coroutines.resume
+import kotlin.coroutines.resumeWithException
+import kotlin.coroutines.suspendCoroutine
+
+private val LOG = getLogger(TorManager::class.java)
+private val EVENTS = listOf(
+    EVENT_CIRCUIT_STATUS, // this one is needed for TorService to function
```

Comment:
Curious: do we still need this if TorService is now using a separate control connection?

Reply:
As there will be two separate control connections then, I think we won't need this any more here. I'll add a TODO.

Reply:
I was getting strange stalls after removing this. The docs aren't totally clear, but this might affect all clients:

https://torproject.gitlab.io/torspec/control-spec/#setevents

Reply:
Weird. Maybe we should do an experiment to check whether it affects all clients? Wouldn't want to start losing events because the TorService control connection called SETEVENTS.

Reply:
It might also *just* be another bug in the TorService library (or the way we use it), because I just saw that again. It just seems to hang, no more events are reported over the control connection.